src/abi/substrate.rs

```diff
@@ -1,284 +1,459 @@
+/// Recoursively build the storage layout after all types are registered
+fn type_to_storage_layout(
+    key: u32,
+    root: &LayoutKey,
+    registry: &PortableRegistryBuilder,
+) -> Layout<PortableForm> {
+    let ty = registry.get(key).unwrap();
+    match ty.type_def() {
+        TypeDef::Composite(inner) => Layout::Struct(StructLayout::new(
+            ty.path().ident().unwrap_or_default(),
+            inner.fields().iter().map(|field| {
+                FieldLayout::new(
+                    field.name().map(ToString::to_string).unwrap_or_default(),
+                    type_to_storage_layout(field.ty().id(), root, registry),
+                )
+            }),
+        )),
+        _ => Layout::Leaf(LeafLayout::new(*root, key.into())),
     }
 }
 
-#[derive(Deserialize, Serialize)]
-pub struct Event {
-    docs: Vec<String>,
-    name: String,
-    args: Vec<ParamIndexed>,
-}
-
-#[derive(Deserialize, Serialize)]
-pub struct Spec {
-    pub constructors: Vec<Constructor>,
-    pub messages: Vec<Message>,
-    pub events: Vec<Event>,
-}
-
-#[derive(Deserialize, Serialize)]
-struct Param {
-    name: String,
-    #[serde(rename = "type")]
-    ty: ParamType,
-}
-
-#[derive(Deserialize, Serialize)]
-struct ParamIndexed {
-    #[serde(flatten)]
-    param: Param,
-    indexed: bool,
-}
+/// Generate `InkProject` from `ast::Type` and `ast::Namespace`
+pub fn gen_project(contract_no: usize, ns: &ast::Namespace) -> InkProject {
+    let mut registry = PortableRegistryBuilder::new();
 
-#[derive(Deserialize, Serialize)]
-struct ParamType {
-    #[serde(rename = "type")]
-    ty: usize,
-    display_name: Vec<String>,
-}
+    // This is only used by off-chain tooling. At the moment there is no such tooling available yet.
+    // So it is not exactly clear yet what this should look like.
+    // For now it just contains all root layouts (you get all storage keys in use).
+    let fields: Vec<FieldLayout<PortableForm>> = ns.contracts[contract_no]
+        .layout
+        .iter()
+        .filter_map(|layout| {
+            let var = &ns.contracts[layout.contract_no].variables[layout.var_no];
+            if let Some(slot) = layout.slot.to_u32() {
+                let ty = resolve_ast(&layout.ty, ns, &mut registry);
+                let layout_key = LayoutKey::new(slot);
+                let root = RootLayout::new(
+                    layout_key,
+                    type_to_storage_layout(ty, &layout_key, &registry),
+                );
+                Some(FieldLayout::new(var.name.clone(), root))
+            } else {
+                None
+            }
+        })
+        .collect();
+    let contract_name = ns.contracts[contract_no].name.clone();
+    let storage = Layout::Struct(StructLayout::new(contract_name, fields));
 
-#[derive(Deserialize, Serialize)]
-struct Storage {
-    #[serde(rename = "struct")]
-    structs: StorageStruct,
-}
+    let f_to_constructor = |f: &Function| -> ConstructorSpec<PortableForm> {
```

Comment:
`f_to_constructor` is not great. How about `function_to_constructor` or simply `constructor_spec`